### DISCUSSION TITLE
Java - 2 - 호제, 진호

### DISCUSSION THREAD

**Discussion by zini9188** [Shortudy-10th/tech-for-developer#20] [질의 응답]
자바 2일차

**Comment by zini9188** [comment]
### ❓️ 추상 클래스와 인터페이스의 차이점에 대해서 말씀해주세요.

<br>

### 💬 답변
<blockquote>
추상 클래스는 extends 키워드를 통해 하위 클래스에서 상속을 받을 수 있지만, 인터페이스는 하위 클래스에서 구현하는 경우 implements 키워드를 사용한다.
추상 클래스는 abstract 키워드를 붙이거나 하나 이상의 추상 메서드로 이루어져 있지만, 인터페이스는 모든 메서드가 추상 메서드이다.
추상 클래스는 다중 상속이 불가능하지만, 인터페이스는 하위 인터페이스나 구현할 클래스에서 다중 상속 및 구현이 가능하다.

</blockquote>

<br>

**Comment by zini9188** [comment]
### ❓️ 객체 지향 프로그래밍의 4가지 특징에 대해서 말씀해주세요.

<br>

### 💬 답변
<blockquote>
객체 지향 프로그래밍의 특징으로는 캡슐화, 추상화, 다형성, 상속이 있다.
캡슐화란 데이터 은닉을 위해 사용하며, 데이터의 구조와 역할, 기능을 하나의 캡슐 형태로 만드는 방법이다.
추상화란 클래스들의 공통적인 특성을 묶어 표현하는 방법이다.
상속이란 상위 클래스의 특징을 하위 클래스에 물려주어 재사용할 수 있는 방법을 의미한다.
다형성이란 상속과 연관이 있는 개념으로 한 객체가 여러 형태로 재구성되는 것을 의미한다.

</blockquote>

<br>

**Comment by zini9188** [comment]
### ❓️ 오버라이딩과 오버로딩에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

오버라이딩은 상속 받는 상위 클래스의 메서드를 재정의하여 하위 클래스에서 사용하는 것이며 해당 메서드의 이름, 매개변수, 반환 타입이 같아야 한다.
오버로딩은 같은 이름과 같은 타입을 반환하는 메서드의 매개변수만 다르게 여러 개 정의하는 것을 의미한다.

</blockquote>

<br>

**Comment by zini9188** [comment]
### ❓️ String, StringBuilder, StringBuffer에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

**String**
- String은 불변(객체의 값은 변경할 수 없다.) 
- '+' 연산을 했을 때 메모리에 새로 + 연산한 값을 저장한 영역을 따로 만들고 변수를 다시 참조하는 식으로 작동한다.
기존에 사용하던 영역은 GC의 제거 대상이 된다.
- 단순한 조회 연산에서는 buffer와 builder 보다 빠르다.

**StringBuffer, StringBuilder**
- StringBuffer와 StringBuilder는 가변(객체의 공간이 부족해지는 경우 버퍼의 크기를 유연하게 늘려준다.)
- 값이 변경될 때마다 새로운 객체를 만드는 String 보다 훨씬 빠르기 때문에, 문자열의 추가, 수정, 삭제가 빈번하게 발생할 경우라면 buffer와 builder를 사용하는 것이 좋다.
- StringBuffer와 StringBuilder의 차이점은 멀티 쓰레드 환경에서 safe한 것(Buffer)과 unsafe(Builder)한 것. Buffer는 동기화(synchronized)를 지원한다.

</blockquote>

<br>

**Reply by zini9188** [reply]
멀티 쓰레드 환경에서 safe하고 unsafe하다는 것의 의미가 무엇인가요?

**Reply by Jihn0118** [reply]
StringBuilder는 동기화를 지원하지 않는 반면, StringBuffer는 동기화를 지원하여 멀티 쓰레드 환경에서도 안전하게 동작할 수 있습니다.
StringBuffer는 메서드에서 synchronized 키워드를 사용하기 때문

**Comment by zini9188** [comment]
### ❓️ 자바에서 스레드를 사용하기 위한 두 가지 방법에 대해 설명하고 차이점을 알려주세요.

<br>

### 💬 답변
<blockquote>

**Thread**
첫번째 방법으로 Thread.
Thread는 쓰레드 생성을 위해 Java에서 미리 구현해둔 클래스이다. 
Thread 클래스로 쓰레드를 구현하려면 이를 상속받는 클래스를 만들고, 내부에서 run()를 구현해야 한다. 그리고 start()를 호출하면 run()이 실행된다.

**Runnable**
두번째 방법은 Runnable 인터페이스를 implements하는 방법이다. Runnbale 인터페이스는 1개의 메소드 (run)만을 갖는 함수형 인터페이스이다. 그렇기 때문에 람다로도 사용 가능하다. 
 
보통 두 번째 방법(Runnable)을 선호한다.
Java의 extends의 상속 방식은 하나의 Class만 가능하기 때문에, 만약 Thread를 사용하기 위해 이 한 번의 상속을 사용한다면 Java의 장점인 상속 기능의 제한이 생기기 때문이다.
그리고 Thread 클래스를 상속받으면 Thread 클래스에 구현된 코드들에 의해 더 많은 자원(메모리, 시간 등)을 필요로 한다.
</blockquote>

<br>

**Reply by zini9188** [reply]
자바에서 스레드를 사용하며 발생할 수 있는 문제와 해결하는 방법을 설명해주세요.

**Reply by Jihn0118** [reply]
스레드를 사용하며 멀티 스레드 환경에서 발생할 수 있는 문제는 Race Condition이 있는데, 
Race Condition은 여러 개의 스레드가 하나의 자원을 놓고 서로 사용하려고 경쟁하는 상황 
 -> Deadlock 또는 Starvation 문제가 발생할 수 있음

- 해결 방법
  - synchronized: synchronized 키워드를 통해 해당 블럭의 액세스를 동기화 할 수 있다.
  - volatile: volatile 키워드를 통해 동일 시점에 모든 스레드가 동일한 값을 가지도록 동기화한다. 하지만 volatile을 통해 모든 동기화 문제가 해결되진 않는다.
 ex) ++ 연산과 같이 원자성이 보장되지 않는 경우 동시성 문제는 다시 발생한다.

  - Atomic 클래스: 위에 두 키워드만으로는 동시성 문제를 깔끔하게 해결할 수 없는데, 자바에서는 위 문제들을 해결하기 위해 비원자 연산에서도 동기화를 빠르고 쉽게 이용할 수 있게 해주는 클래스 모음을 제공함(java.util.concurrent.*)
